### ISSUE TITLE
Add an option to strip query string from URLs

### ISSUE THREAD

**Issue by dpyryesk** [yasserg/crawler4j#81]
Sometimes you do not want to crawl the same page multiple times because other pages are leading to it with various query strings. For example: http://somesite.com/search?query=foo, http://somesite.com/search?query=bar and so on.

**Comment by Chaiavi** [comment]
The best option for handling this will be to override:
protected WebURL handleUrlBeforeProcess(WebURL curURL) {
    return curURL;
  }

In your crawler class (the class which extends WebCrawler).


Any other generic way doesn't make sense.

**Comment by dpyryesk** [comment]
I tried that. If you tweak the URL in such a way, then you get several of the same ones in the list. Of course I can clean up the list, but processing time is still wasted on looking at the same page over and over.

**Comment by Chaiavi** [comment]
The waste is really really minor.  That stage takes milliseconds.

**Comment by dpyryesk** [comment]
It still downloads and parses the page multiple times without any point. It also puts unnecessary stress on the website being crawled. In my test the same page was downloaded at least 10 times.

**Comment by Chaiavi** [comment]
What about the shouldVisit() method ?


You can dictate your restriction there, that one happens before the actual
downloading of the page

On Fri, Jul 10, 2015 at 12:24 AM, Dmitry Pyryeskin <notifications@github.com

**Comment by dpyryesk** [comment]
I do not think that crawlers have the access to the database of links, so they can't verify if the link is duplicate or not. If I skip all links with query string, then I might not crawl some of the pages at all.
I still think that Parser should accept an option to strip the query string and duplicate pages should be handled by Frontier.

**Comment by Chaiavi** [comment]
Ahh, I see.


I have just checked the code.

It seems that we don't have a hook for that.


As I see it, you can do one of the following:
1. Suggest your patch by using a pull request and if the project owner thinks it is good he will merge.
2. Use handleUrlBeforeProcess() and there throw an exception.
3. Use the shouldVisit() - where you would keep an internal list against which you can compare the urls.
